src/generic_mtrie.hpp

```diff
@@ -0,0 +1,111 @@
+/*
+Copyright (c) 2018 Contributors as noted in the AUTHORS file
+
+This file is part of libzmq, the ZeroMQ core engine in C++.
+
+libzmq is free software; you can redistribute it and/or modify it under
+the terms of the GNU Lesser General Public License (LGPL) as published
+by the Free Software Foundation; either version 3 of the License, or
+(at your option) any later version.
+
+As a special exception, the Contributors give you permission to link
+this library with independent modules to produce an executable,
+regardless of the license terms of these independent modules, and to
+copy and distribute the resulting executable under terms of your choice,
+provided that you also meet, for each linked independent module, the
+terms and conditions of the license of that module. An independent
+module is a module which is not derived from or based on this library.
+If you modify this library, you must extend this exception to your
+version of the library.
+
+libzmq is distributed in the hope that it will be useful, but WITHOUT
+ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
+License for more details.
+
+You should have received a copy of the GNU Lesser General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#ifndef __ZMQ_GENERIC_MTRIE_HPP_INCLUDED__
+#define __ZMQ_GENERIC_MTRIE_HPP_INCLUDED__
+
+#include <stddef.h>
+#include <set>
+
+#include "stdint.hpp"
+
+namespace zmq
+{
+//  Multi-trie. Each node in the trie is a set of pointers to pipes.
+template <typename T> class generic_mtrie_t
+{
+  public:
+    typedef T value_t;
+    typedef const unsigned char *prefix_t;
+
+    generic_mtrie_t ();
+    ~generic_mtrie_t ();
+
+    //  Add key to the trie. Returns true if it's a new subscription
+    //  rather than a duplicate (i.e. an entry with the same prefix
+    //  but a different pipe already exists).
+    //  TODO what if this is called with the same prefix AND pipe?
+    //  Is this legal? It is not checked anywhere.
```

Comment:
To subscribe twice to the same topic on the same socket doesn't make much sense, but it is never specified that duplicated subscriptions are illegal so we can't raise errors, and I guess it kinda makes sense that these duplicates are just ignored.
As far as I can see internally it's a harmless no-op, given the pipes are stored in a set (which I suspect was intentional, to avoid these kind of dups)

Reply:
In other words, from the point of view of the implementation, all that we care about when inserting a new sub is whether there were other sockets already subscribed or not (it's an XPUB feature) - and the code already does the right thing regardless of whether the other subscribers are the same or different sockets (pipes).

Reply:
Furthermore, XSUB does check for duplication but intentionally sends the duplicated subscription messages to the XPUB anyway: https://github.com/zeromq/libzmq/blob/master/src/xsub.cpp#L101

Reply:
I am not 100% sure how to map this to the mtrie_t. Are you saying that calling add should be idempotent, i.e. if you call it multiple times with the same arguments, it should behave as if called only once? I don't think this is the current behaviour, probably it would add multiple equivalent entries, but I need to check that, and add a test case.

Reply:
Isn't that what the set data structure is for? If the same pipe is already in the set, adding it again won't do anything, right?

Reply:
Yes, it is a set, forget about that. I added a test case, and it indeed behaves as you expected.

Reply:
@sigiesec I was reading the RFCs and I remembered this discussion - turns out the spec says that subscribing twice to A means that you also need to unsubscribe twice:

> Subscriptions SHALL be additive and SHALL NOT be idempotent. That is, subscribing to "A" and "" is the same as subscribing to "" alone. Subscribing to "A" and "A" counts as two subscriptions, and would require two unsubscribe messages to undo.

So duplicated subs are really not illegal and there's an expected behaviour. The tests that were added here exercise that behaviour - but I'm worried that it's implemented in the client side, by the XSUB mtrie.

What did we say happens right now if a duplicated subscription is actually received by the PUB? I'm thinking about cross-implementation compatibility specifically

Reply:
Basically I'm worried that on the client side we are fine as we just ignore them, but on the PUB side we are not doing the right thing given a set is used.